src/serac/numerics/functional/functional.hpp

```diff
@@ -69,66 +109,77 @@ class Functional;
  * my_residual.AddDomainIntegral(Dimension<3>{}, integrand, domain_of_integration);
  * @endcode
  */
-template <typename test, typename trial, ExecutionSpace exec>
-class Functional<test(trial), exec> : public mfem::Operator {
+template <typename test, typename... trials, ExecutionSpace exec>
+class Functional<test(trials...), exec> {
+  static constexpr tuple<trials...> trial_spaces{};
+  static constexpr uint32_t         num_trial_spaces = sizeof...(trials);
+
+  class Gradient;
+
+  template <typename... T>
+  struct operator_paren_return {
+    using type = typename std::conditional<
+        (std::is_same_v<T, dual_vector> + ...) == 1,  // if the there is a dual number in the pack
+        serac::tuple<mfem::Vector&, Gradient&>,       // then we return the value and the derivative
+        mfem::Vector&                                 // otherwise, we just return the value
+        >::type;
+  };
+
 public:
   /**
    * @brief Constructs using @p mfem::ParFiniteElementSpace objects corresponding to the test/trial spaces
    * @param[in] test_fes The (non-qoi) test space
    * @param[in] trial_fes The trial space
    */
-  Functional(mfem::ParFiniteElementSpace* test_fes, mfem::ParFiniteElementSpace* trial_fes)
-      : Operator(test_fes->GetTrueVSize(), trial_fes->GetTrueVSize()),
-        test_space_(test_fes),
-        trial_space_(trial_fes),
-        P_test_(test_space_->GetProlongationMatrix()),
-        G_test_(test_space_->GetElementRestriction(mfem::ElementDofOrdering::LEXICOGRAPHIC)),
-        P_trial_(trial_space_->GetProlongationMatrix()),
-        G_trial_(trial_space_->GetElementRestriction(mfem::ElementDofOrdering::LEXICOGRAPHIC)),
-        grad_(*this)
+  Functional(mfem::ParFiniteElementSpace*                               test_fes,
+             std::array<mfem::ParFiniteElementSpace*, num_trial_spaces> trial_fes)
+      : test_space_(test_fes), trial_space_(trial_fes)
   {
-    SLIC_ERROR_IF(!G_test_, "Couldn't retrieve element restriction operator for test space");
-    SLIC_ERROR_IF(!G_trial_, "Couldn't retrieve element restriction operator for trial space");
+    for (uint32_t i = 0; i < num_trial_spaces; i++) {
+      P_trial_[i] = trial_space_[i]->GetProlongationMatrix();
+      G_trial_[i] = trial_space_[i]->GetElementRestriction(mfem::ElementDofOrdering::LEXICOGRAPHIC);
+      SLIC_ERROR_IF(!G_trial_[i], "Couldn't retrieve element restriction operator for trial space");
 
-    // Ensure the mesh has the appropriate neighbor information before constructing the face restriction operators
-    if (test_space_) {
-      test_space_->ExchangeFaceNbrData();
-    }
-    if (trial_space_) {
-      trial_space_->ExchangeFaceNbrData();
+      if (compatibleWithFaceRestriction(*trial_space_[i])) {
```

Comment:
What happens if a fe space is not compatible with face restrictions? I assume it generates and error when the boundary integral is requested.